File: Chapter03/3.3.4 ISP/BadWay/Midfielder.cs
namespace Isp.BadWay
{
    public class Midfielder : ITrainingSession
    {
        public void PracticeShooting() { Console.WriteLine("Practicing shooting..."); }
        public void PracticeTackling() { Console.WriteLine("Practicing tackling..."); }

        public void PracticeDivingSaves()
        {
            throw new NotImplementedException("I'm a midfielder, not a goalie!");
        }
        public void PracticeHandDistribution()
        {
            throw new NotImplementedException("I'm a midfielder, not a goalie!");
        }
    }
}

File: Chapter 03/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter03/3.3.4 ISP/Before/Midfielder.java
package com.grokkingsoftwarearchitecture.chapter03.isp.before;

public class Midfielder implements TrainingSession {
    public void practiceShooting() {
        System.out.println("  [Midfielder] Practicing shooting drills.");
    }

    public void practiceTackling() {
        System.out.println("  [Midfielder] Practicing slide tackles.");
    }

    public void practiceDivingSaves() {
        throw new UnsupportedOperationException("Midfielders don't play in the net!");
    }

    public void practiceHandDistribution() {
        throw new UnsupportedOperationException("Midfielders can't use their hands!");
    }
}